File: src/main/java/emufog/export/MaxiNetExporter.java
/*
 * MIT License
 *
 * Copyright (c) 2018 emufog contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package emufog.export;

import emufog.docker.DockerType;
import emufog.graph.Edge;
import emufog.graph.EmulationSettings;
import emufog.graph.Graph;
import emufog.graph.Node;
import emufog.settings.Settings;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * This class exports a graph object to a valid python file usable with the
 * MaxiNet (https://maxinet.github.io/) network emulation framework.
 */
public class MaxiNetExporter implements IGraphExporter {

    /* list of all lines of the respective file in top down order */
    private final List<String> lines;

    /* blank line object to reuse for all blank lines */
    private final String blankLine;

    /* mapping of edges to their respective connector */
    private final Map<Edge, String> connectors;

    /**
     * Creates a new MaxiNet exporter instance.
     */
    public MaxiNetExporter() {
        lines = new ArrayList<>();
        blankLine = "";
        connectors = new HashMap<>();
    }

    @Override
    public void exportGraph(Graph graph, Path path) throws IllegalArgumentException, IOException {
        if (graph == null) {
            throw new IllegalArgumentException("The given graph object does not exist.");
        }
        if (path == null) {
            throw new IllegalArgumentException("The given path is null. Please provide a valid path");
        }

        // check if file exists and can be overwritten
        Settings settings = graph.getSettings();
        File file = path.toFile();
        if (!settings.overwriteExperimentFile && file.exists()) {
            throw new IllegalArgumentException("The given file already exist. Please provide a valid path");
        }

        // check the file ending of the given path
        PathMatcher matcher = FileSystems.getDefault().getPathMatcher("glob:**.py");
        if (!matcher.matches(path)) {
            throw new IllegalArgumentException("The file name for MaxiNet hat to be a python file (.py)");
        }

        // initialize empty sets to start the writing
        lines.clear();
        connectors.clear();

        // begin to write the python file
        setupImports();

        addBlankLine();
        lines.add("topo = Topo()");
        addHosts(graph);
        addSwitches(graph);
        addConnectors(graph);
        addLinks(graph);
        setupExperiment();

        // set the overwrite option if feature is set in the settings file
        StandardOpenOption overwrite = settings.overwriteExperimentFile ? StandardOpenOption.TRUNCATE_EXISTING : StandardOpenOption.APPEND;
        // write output in UTF-8 to the specified file
        Files.write(file.toPath(), lines, StandardCharsets.UTF_8, StandardOpenOption.CREATE, overwrite);
    }

    /**
     * Adds a blank line to the output file.
     */
    private void addBlankLine() {
        lines.add(blankLine);
    }

    /**
     * Writes all docker host nodes of the graph to the output file.
     *
     * @param graph graph to export
     */
    private void addHosts(Graph graph) {
        addBlankLine();
        lines.add("# add hosts");

        for (Node n : graph.getNodes().stream().filter(Node::hasEmulationSettings).collect(Collectors.toList())) {
            EmulationSettings emu = n.getEmulationNode();
            DockerType docker = emu.getDockerType();
            lines.add(n.getName() + " = topo.addHost(\"" + n.getName() + "\", cls=Docker, ip=\"" + emu.getIP() +
                    "\", dimage=\"" + docker.dockerImage + "\", mem_limit=" + docker.memoryLimit + ")");
        }
    }

    /**
     * Writes all switches that do not require docker to the output file.
     *
     * @param graph graph to export
     */
    private void addSwitches(Graph graph) {
        addBlankLine();
        lines.add("# add switches");

        List<Node> nodes = new ArrayList<>();
        nodes.addAll(graph.getRouters());
        nodes.addAll(graph.getSwitches());
        for (Node n : nodes.stream().filter(n -> !n.hasEmulationSettings()).collect(Collectors.toList())) {
            lines.add(n.getName() + " = topo.addSwitch(\"" + n.getName() + "\")");
        }
    }

    /**
     * Creates connectors between two hosts to run in MaxiNet.
     *
     * @param graph graph to export
     */
    private void addConnectors(Graph graph) {
        addBlankLine();
        lines.add("# add connectors");

        int counter = 0;
        for (Edge e : graph.getEdges()) {
            if (e.getSource().hasEmulationSettings() && e.getDestination().hasEmulationSettings()) {
                String name = "c" + counter;
                lines.add(name + " = topo.addSwitch(\"" + name + "\")");
                connectors.put(e, name);
                counter++;
            }
        }
    }

    /**
     * Established the connections between two nodes based on the edges of the graph.
     *
     * @param graph graph to export
     */
    private void addLinks(Graph graph) {
        addBlankLine();
        lines.add("# add links");

        for (Edge e : graph.getEdges()) {
            if (connectors.containsKey(e)) {
                String connector = connectors.get(e);
                addLink(e.getSource().getName(), connector, e.getDelay() / 2, e.getBandwidth());
                addLink(connector, e.getDestination().getName(), e.getDelay() / 2, e.getBandwidth());
            } else {
                addLink(e.getSource().getName(), e.getDestination().getName(), e.getDelay(), e.getBandwidth());
            }
        }
    }

    /**
     * Adds a new link between two nodes to the document.
     *
     * @param source      source of the link
     * @param destination destination of the link
     * @param latency     latency applied to this link
     * @param bandwidth   bandwidth limitations of this link
     */
    private void addLink(String source, String destination, float latency, float bandwidth) {
        lines.add("topo.addLink(" + source + ", " + destination +
                ", delay='" + latency + "ms', bw=" + bandwidth + ")");
    }

    /**
     * Writes the necessary imports at the top of the output file.
     */
    private void setupImports() {
        lines.add("#!/usr/bin/env python2");
        addBlankLine();
        lines.add("import time");
        addBlankLine();
        lines.add("from MaxiNet.Frontend import maxinet");
        lines.add("from MaxiNet.Frontend.container import Docker");
        lines.add("from mininet.topo import Topo");
        lines.add("from mininet.node import OVSSwitch");
    }

    /**
     * Writes the lines to setup and start an experiment in MaxiNet.
     */
    private void setupExperiment() {
        addBlankLine();
        lines.add("# create experiment");
        lines.add("cluster = maxinet.Cluster()");
        lines.add("exp = maxinet.Experiment(cluster, topo, switch=OVSSwitch)");
        lines.add("exp.setup()");
    }
}

File: src/main/kotlin/emufog/export/maxinet/MaxiNetExporter.kt
/*
 * MIT License
 *
 * Copyright (c) 2018 emufog contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package emufog.export.maxinet

import emufog.export.GraphExporter
import emufog.graph.Edge
import emufog.graph.Graph
import emufog.graph.Node
import emufog.graph.NodeType.BACKBONE_NODE
import emufog.graph.NodeType.EDGE_DEVICE_NODE
import emufog.graph.NodeType.EDGE_NODE
import java.io.BufferedWriter
import java.nio.file.FileSystems
import java.nio.file.Path

/**
 * This object exports a graph object to a valid python file usable with the [MaxiNet](https://maxinet.github.io/)
 * network emulation framework.
 */
object MaxiNetExporter : GraphExporter {

    override fun exportGraph(graph: Graph, path: Path) {
        MaxiNetExporterImpl(graph, path).exportGraph()
    }
}

private class MaxiNetExporterImpl internal constructor(private val graph: Graph, path: Path) {

    private val writer: BufferedWriter

    /**
     * mapping of edges to their respective connector
     */
    private val connectors: MutableMap<Edge, String> = HashMap()

    init {
        // check if file exists and can be overwritten
        val config = graph.config
        val file = path.toFile()
        require(!(!config.overWriteOutputFile && file.exists())) {
            "The given file already exist. Please provide a valid path."
        }

        writer = path.toFile().bufferedWriter()

        // check the file ending of the given path
        val matcher = FileSystems.getDefault().getPathMatcher("glob:**.py")
        require(matcher.matches(path)) { "The file name for MaxiNet has to be a python file (.py)." }
    }

    private fun BufferedWriter.writeln(s: String) {
        this.write(s)
        this.newLine()
    }

    internal fun exportGraph() {
        // begin to write the python file
        writer.writeln("#!/usr/bin/env python2")
        writer.newLine()
        writer.writeln("import time")
        writer.newLine()
        writer.writeln("from MaxiNet.Frontend import maxinet")
        writer.writeln("from MaxiNet.Frontend.container import Docker")
        writer.writeln("from mininet.topo import Topo")
        writer.writeln("from mininet.node import OVSSwitch")
        writer.newLine()
        writer.writeln("topo = Topo()")
        writer.newLine()
        addHosts()
        addSwitches()
        addConnectors()
        addLinks()
        writer.newLine()
        writer.writeln("# create experiment")
        writer.writeln("cluster = maxinet.Cluster()")
        writer.writeln("exp = maxinet.Experiment(cluster, topo, switch=OVSSwitch)")
        writer.writeln("exp.setup()")

        writer.flush()
        writer.close()
    }

    /**
     * Writes all container host nodes of the graph to the output file.
     */
    private fun addHosts() {
        writer.writeln("# add hosts")
        graph.nodes
            .filter { it.hasEmulationSettings() }
            .forEach {
                val emu = it.emulationNode!!
                val container = emu.container
                writer.writeln(
                    String.format(
                        "%s = topo.addHost(\"%s\", cls=Docker, ip=\"%s\", dimage=\"%s\", mem_limit=%d)",
                        getName(it),
                        getName(it),
                        emu.ip,
                        container.fullName(),
                        container.memoryLimit
                    )
                )
            }
    }

    /**
     * Writes all switches that do not require container to the output file.
     */
    private fun addSwitches() {
        writer.newLine()
        writer.writeln("# add switches")
        graph.edgeNodes.union(graph.backboneNodes)
            .filter { !it.hasEmulationSettings() }
            .forEach { writer.writeln(String.format("%s = topo.addSwitch(\"%s\")", getName(it), getName(it))) }
    }

    /**
     * Creates connectors between two hosts to run in MaxiNet.
     */
    private fun addConnectors() {
        writer.writeln("")
        writer.writeln("# add connectors")
        var counter = 0
        graph.edges
            .filter { it.source.hasEmulationSettings() && it.destination.hasEmulationSettings() }
            .forEach {
                val name = "c$counter"
                writer.writeln(String.format("%s = topo.addSwitch(\"%s\")", name, name))
                connectors[it] = name
                counter++
            }
    }

    /**
     * Established the connections between two nodes based on the edges of the graph.
     */
    private fun addLinks() {
        writer.writeln("")
        writer.writeln("# add links")
        graph.edges.forEach {
            if (connectors.containsKey(it)) {
                val connector = connectors[it]
                addLink(getName(it.source), connector, it.latency / 2, it.bandwidth)
                addLink(connector, getName(it.destination), it.latency / 2, it.bandwidth)
            } else {
                addLink(getName(it.source), getName(it.destination), it.latency, it.bandwidth)
            }
        }
    }

    /**
     * Adds a new link between two nodes to the document.
     *
     * @param source source of the link
     * @param destination destination of the link
     * @param latency latency applied to this link
     * @param bandwidth bandwidth limitations of this link
     */
    private fun addLink(source: String?, destination: String?, latency: Float, bandwidth: Float) {
        writer.writeln(
            String.format("topo.addLink(%s, %s, delay='%fms', bw=%f)", source, destination, latency, bandwidth)
        )
    }

    private fun getName(node: Node): String {
        val prefix = when (node.type) {
            BACKBONE_NODE -> "s"
            EDGE_NODE -> "r"
            EDGE_DEVICE_NODE -> "h"
        }

        return prefix + node.id
    }
}